instance/models/database_server.py

```diff
@@ -198,13 +213,112 @@ def default_port(self):
         return MYSQL_SERVER_DEFAULT_PORT
 
 
+class MongoDBReplicaSetManager(models.Manager):
+    """
+    Custom manager for the DatabaseServer model.
+    """
+    def _get_setting(self, field):
+        """
+        Get setting using prefix
+        """
+        return getattr(
+            settings,
+            self.model.DEFAULT_SETTINGS_NAME + '_' + field,
+            None
+        )
+
+    def get_replica_set_settings(self):
+        """
+        Create dictionary with replica settings
+        """
+        return {
+            'user': self._get_setting('USER'),
+            'password': self._get_setting('PASSWORD'),
+            'name': self._get_setting('NAME'),
+            'primary': self._get_setting('PRIMARY'),
+            'hosts': self._get_setting('HOSTS'),
+            'port': self._get_setting('PORT')
+        }
+
+    def _create_default(self):
+        """
+        Create the default database replica set configured in the Django settings, if any.
+        """
+        replica_settings = self.get_replica_set_settings()
+        for setting in replica_settings:
+            if setting != 'port' and replica_settings[setting] is None:
+                logger.error(
+                    "Error creating the default servers for the replica set, please ensure that"
+                    " all needed settings are configured."
+                )
+                return
+
+        replica_set_hosts = replica_settings['hosts'].split(',')
+        replica_set, _ = self.get_or_create(name=replica_settings['name'])
+        for host in replica_set_hosts:
+            MongoDBServer.objects.create_database_server(
+                host=host,
+                user=replica_settings['user'],
+                password=replica_settings['password'],
+                port=replica_settings['port'],
+                primary=replica_settings['primary'],
+                replica_set=replica_set
+            )
+
+    def select_random(self):
+        """
+        Select a replica set for a new instance.
+        The current implementation selects one of the replica sets with servers that accept new clients at random.
+        If no database server accepts new clients, DoesNotExist is raised.
+        """
+        self._create_default()
+
+        # The set of servers might change between retrieving the server count and retrieving the random server,
+        # so we make this atomic.
+        with transaction.atomic():
```

Comment:
@rocioar I understand this was migrated from previous version, but can't we simplify it to not require atomic transaction?

Looks like something like this might just do the same:

1. Fetch all distinct mongoDB servers **replica sets**
2. Count them
3. Do the "count > 0" check
4. Return random replica set.

The difference is that we fetch replica sets right away, and than just pick one random object from "memory", while current implementation fetches objects, than picks one random from them, than goes to DB to fetch a field.

Reply:
@e-kolpakov This function is also checking that the MongoDB severs are accepting clients, do you think it would be a good idea to avoid that check?

Also to avoid the extra query I can add a select_related